File: FinalProject_OOP_Java/src/Guard/Guard.java
package Guard;

import System.SystemPark;
import Time.TimeSpan;
import Vehicle.Vehicle;
import Vehicle.Vehicle_List;

public class Guard implements Guard_Data_Cal,Guard_Vehicle_OUT
{
	//Các thuộc tính
	private String Gua_sID;
	private String Gua_sName;
	private int Gua_iCa;
	private long Gua_lLuongCB;
	private TimeSpan Gua_tTime_lam_them = new TimeSpan();
	private String Gua_sAddress;
	private String Gua_sCall_number;
	
	//Các định nghĩa thuộc tính
	public String getGua_sID() {
		return Gua_sID;
	}
	public void setGua_sID(String gua_sID) {
		Gua_sID = gua_sID;
	}
	public String getGua_sName() {
		return Gua_sName;
	}
	public void setGua_sName(String gua_sName) {
		Gua_sName = gua_sName;
	}
	public int getGua_iCa() {
		return Gua_iCa;
	}
	public void setGua_iCa(int gua_iCa) {
		if (gua_iCa > 3 || gua_iCa <= 0)
            throw new ArithmeticException("EROLL");
		Gua_iCa = gua_iCa;
	}
	public long getGua_lLuongCB() {
		return Gua_lLuongCB;
	}
	public void setGua_lLuongCB(long gua_lLuongCB) {
        if (gua_lLuongCB < 0)
            throw new ArithmeticException("EROLL");
        this.Gua_lLuongCB = gua_lLuongCB;
		Gua_lLuongCB = gua_lLuongCB;
	}
	public TimeSpan getGua_tTime_lam_them() {
		return Gua_tTime_lam_them;
	}
	public void setGua_tTime_lam_them(TimeSpan gua_tTime_lam_them) {
		Gua_tTime_lam_them = gua_tTime_lam_them;
	}
	public String getGua_sAddress() {
		return Gua_sAddress;
	}
	public void setGua_sAddress(String gua_sAddress) {
		Gua_sAddress = gua_sAddress;
	}
	public String getGua_sCall_number() {
		return Gua_sCall_number;
	}
	public void setGua_sCall_number(String gua_sCall_number) {
		Gua_sCall_number = gua_sCall_number;
	}
	
	//Khởi tạo
	public Guard()
    {
		super();
    }

    public Guard(String id, String name, int ca, long luongcb, TimeSpan time_them, String address, String call_num)
    {
    	super();
        this.setGua_sID(id);
        this.setGua_sName(name);
        this.setGua_iCa(ca);
        this.setGua_lLuongCB(luongcb);
        this.setGua_tTime_lam_them(time_them);
        this.setGua_sAddress(address);
        this.setGua_sCall_number(call_num);
    }

    public Guard(String id, String name, int ca, String address, String call_num)
    {
    	super();
    	this.setGua_sID(id);
        this.setGua_sName(name);
        this.setGua_iCa(ca);
        this.setGua_sAddress(address);
        this.setGua_sCall_number(call_num);
    }

    public Guard(String id, String name, int ca, String call_num)
    {
    	super();
    	this.setGua_sID(id);
        this.setGua_sName(name);
        this.setGua_iCa(ca);
        this.setGua_sCall_number(call_num);
    }

    public Guard(String id, int ca, long luongcb, TimeSpan time_them, String call_num)
    {
    	super();
    	this.setGua_sID(id);
        this.setGua_iCa(ca);
        this.setGua_lLuongCB(luongcb);
        this.setGua_tTime_lam_them(time_them);
        this.setGua_sCall_number(call_num);
    }

    public Guard(String id, String name, int ca, long luongcb)
    {
    	super();
    	this.setGua_sID(id);
        this.setGua_sName(name);
        this.setGua_iCa(ca);
        this.setGua_lLuongCB(luongcb);
    }
   
	//Các hàm
    public void Gua_Input(String id, String name, int ca, long luongcb,TimeSpan time_them, String address, String call_num)
    {
        this.setGua_sID(id);
        this.setGua_sName(name);
        this.setGua_iCa(ca);
        this.setGua_lLuongCB(luongcb);
        this.setGua_tTime_lam_them(time_them);
        this.setGua_sAddress(address);
        this.setGua_sCall_number(call_num);
    }
    
    public void Gua_Output()
    {
    	System.out.printf("|%23s|%20s|%6s|%11s|%9s|%25s|%15s|\n",this.Gua_sID , this.Gua_sName ,this.Gua_iCa , this.Gua_Calculate_Salary_1day() , this.Gua_tTime_lam_them.ToString() , this.Gua_sAddress ,this.Gua_sCall_number);
      //  System.out.println("ID: " + this.Gua_sID);
      //  System.out.println("Name: " + this.Gua_sName);
      //  System.out.println("Ca: " + this.Gua_iCa);
      //  System.out.println("Luong CB: " + this.Gua_lLuongCB);
      //  System.out.println("Thoi gian lam them: " + this.Gua_tTime_lam_them.ToString());
      //  System.out.println("Address: " + this.Gua_sAddress);
      //  System.out.println("Call_Number: " + this.Gua_sCall_number);
    }
    //Tính tiền lương /1 ngày
    public double Gua_Calculate_Salary_1day()
    {
        long t = this.Gua_tTime_lam_them.diffHours;
        return 200000 + this.Gua_lLuongCB*t ;
    }
   
    //trả về string khu vực quản lý
    public String Gua_Patrol_area()
    {
        String[] b= this.Gua_sID.split("_");
        return "Khu vuc tuan tra: " + b[3].charAt(0) + " , " + b[3].charAt(1) + " , " + b[3].charAt(2) + " , " + b[3].charAt(3);
    }
    
    //Trả về string thời gian làm việc cố định
    public String Gua_Patrol_time()
    {
        String str = "Thoi gian tuan tra: ";
        if (this.Gua_iCa == 1)
            str += "0h --> 8h";
        else if (this.Gua_iCa == 2)
            str += "8h --> 16h";
        else
            str += "16h --> 0h";
        return str;
    }
    //Các truy cập của Guard
	@Override
	//Xuất 1 bảo vệ
	public void Gua_Output_Person(Guard person) {
		// TODO Auto-generated method stub
		System.out.println("Information of Guard <Person>:");
		System.out.printf("|%23s|%20s|%6s|%11s|%9s|%25s|%15s|\n","ID","Name","Shifts","Salary/1day","Overtime","Address","Call Number");
        person.Gua_Output();
		
	}
	
	@Override
	//Xuất List_Veh_On
	public void Gua_List_Veh_ON() {
		// TODO Auto-generated method stub
		System.out.println("___List of vehicles that are parking___");
        Vehicle_List.OutList(SystemPark.List_Vehicle_On_Main);
		
	}
	@Override
	//Xuất List_Veh_Off
	public void Gua_List_Veh_OFF() {
		// TODO Auto-generated method stub
		System.out.println("___List of vehicles that were parking___");
        Vehicle_List.OutList(SystemPark.List_Vehicle_Off_Main);
	}
	
	 //Tìm và Xóa 1 vehicle trong List_Vehicle_On_Main
    public Vehicle Find_Delete_Vehicle(String veh_ID)
    {
        for (int i = 0; i < SystemPark.List_Vehicle_On_Main.size(); i++)
            if (SystemPark.List_Vehicle_On_Main.get(i).getVeh_sID().compareTo(veh_ID)==0)
            {
            	Vehicle veh= SystemPark.List_Vehicle_On_Main.get(i);
                SystemPark.List_Vehicle_On_Main.remove(i);
                System.out.println("Successful delete Vehicle in List_Vehicle_On!!");
                return veh;
            }
        System.out.println("Delete failed!!");
        return null;
    }
}

File: FinalProject_OOP_C#/Guard.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FinalProject_OOP
{
    public class Guard : Guard_Data_Cal, Guard_Vehicle_OUT
    {
        private string Gua_sID;
        private string Gua_sName;
        private int Gua_iCa;
        private long Gua_lLuongCB;
        private TimeSpan Gua_tTime_lam_them = TimeSpan.Parse("0:0:0:0");
        private string Gua_sAddress;
        private string Gua_sCall_number;

        public string Gua_ID
        {
            get { return this.Gua_sID; }
            set { this.Gua_sID = value; }
        }

        public string Gua_Name
        {
            get { return this.Gua_sName; }
            set { this.Gua_sName = value; }
        }

        public int Gua_Ca
        {
            get { return this.Gua_iCa; }
            set
            {
                if (value > 3 || value <= 0)
                    throw new ArgumentOutOfRangeException(
                         $"{nameof(value)} khong hop le !");
                this.Gua_iCa = value;
            }
        }

        public long Gua_LuongCB
        {
            get { return this.Gua_lLuongCB; }
            set
            {
                if (value < 0)
                    throw new ArgumentOutOfRangeException(
                          $"{nameof(value)} khong hop le !");
                this.Gua_lLuongCB = value;
            }
        }

        public TimeSpan Gua_Time_lam_them
        {
            get { return this.Gua_tTime_lam_them; }
            set { this.Gua_tTime_lam_them = value; }
        }

        public string Gua_Address
        {
            get { return this.Gua_sAddress; }
            set { this.Gua_sAddress = value; }
        }

        public string Gua_Call_number
        {
            get { return this.Gua_sCall_number; }
            set { this.Gua_sCall_number = value; }
        }

        public Guard()
        {
        }

        public Guard(string id, string name, int ca, long luongcb, TimeSpan time_them, string address, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_LuongCB = luongcb;
            this.Gua_Time_lam_them = time_them;
            this.Gua_Address = address;
            this.Gua_Call_number = call_num;
        }

        public Guard(string id, string name, int ca, string address, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_Address = address;
            this.Gua_Call_number = call_num;
        }

        public Guard(string id, string name, int ca, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_Call_number = call_num;
        }

        public Guard(string id, int ca, long luongcb, TimeSpan time_them, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Ca = ca;
            this.Gua_LuongCB = luongcb;
            this.Gua_Time_lam_them = time_them;
            this.Gua_Call_number = call_num;
        }

        public Guard(string id, string name, int ca, long luongcb)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_LuongCB = luongcb;
        }

        ~Guard()
        {
        }

        public void Gua_Input()
        {
            Console.Write("Nhap ID: ");
            this.Gua_ID = Console.ReadLine();
            Console.Write("Nhap Name: ");
            this.Gua_Name = Console.ReadLine();
            Console.Write("Nhap Ca: ");
            this.Gua_Ca = int.Parse(Console.ReadLine());
            Console.Write("Nhap Luong CB: ");
            this.Gua_LuongCB = long.Parse(Console.ReadLine());
            Console.Write("Nhap thoi gian lam them: ");
            this.Gua_Time_lam_them = TimeSpan.Parse(Console.ReadLine());
            Console.Write("Nhap Address: ");
            this.Gua_Address = Console.ReadLine();
            Console.Write("Nhap Call_Number: ");
            this.Gua_Call_number = Console.ReadLine();
        }

        public void Gua_Input(string id, string name, int ca, long luongcb, TimeSpan time_them, string address, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_LuongCB = luongcb;
            this.Gua_Time_lam_them = time_them;
            this.Gua_Address = address;
            this.Gua_Call_number = call_num;
        }

        public void Gua_Input(string id, string name, int ca, string address, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_Address = address;
            this.Gua_Call_number = call_num;
        }

        public void Gua_Input(string id, string name, int ca, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_Call_number = call_num;
        }

        public void Gua_Input(string id, int ca, long luongcb, TimeSpan time_them, string call_num)
        {
            this.Gua_ID = id;
            this.Gua_Ca = ca;
            this.Gua_LuongCB = luongcb;
            this.Gua_Time_lam_them = time_them;
            this.Gua_Call_number = call_num;
        }

        public void Gua_Input(string id, string name, int ca, long luongcb)
        {
            this.Gua_ID = id;
            this.Gua_Name = name;
            this.Gua_Ca = ca;
            this.Gua_LuongCB = luongcb;
        }

        public void Gua_Output()
        {
            Console.WriteLine("ID: " + this.Gua_ID);
            Console.WriteLine("Name: " + this.Gua_Name);
            Console.WriteLine("Ca: " + this.Gua_Ca);
            Console.WriteLine("Luong CB: " + this.Gua_LuongCB);
            Console.WriteLine("Thoi gian lam them: " + this.Gua_Time_lam_them);
            Console.WriteLine("Address: " + this.Gua_Address);
            Console.WriteLine("Call_Number: " + this.Gua_Call_number);
        }
        public void Gua_OutPut()
        {
            Console.WriteLine(this.Gua_ID + "|  " + this.Gua_Name + "  |   " + this.Gua_Ca + "  |     " + this.Gua_Calculate_Salary_1day() + "    | " + this.Gua_Time_lam_them + " |" + this.Gua_Address + "|      "+this.Gua_Call_number+"       |");
        }
        public double Gua_Calculate_Salary_1day()
        {
            double t = this.Gua_Time_lam_them.Days * 24 + this.Gua_Time_lam_them.Hours + this.Gua_Time_lam_them.Minutes / 60 + this.Gua_Time_lam_them.Seconds / (60 * 60);
            return 200000 + 50000 * t;
        }

        public string Gua_Patrol_area()
        {
            string[] b = this.Gua_ID.Split('_');
            return "Khu vuc tuan tra: " + b[3][0] + " , " + b[3][1] + " , " + b[3][2] + " , " + b[3][3];
        }

        public string Gua_Patrol_time()
        {
            string str = "Thoi gian tuan tra: ";
            if (this.Gua_Ca == 1)
                str += "0h --> 8h";
            else if (this.Gua_Ca == 2)
                str += "8h --> 16h";
            else
                str += "16h --> 0h";
            return str;
        }
        public void Gua_List_Veh_ON()
        {
            Console.WriteLine("___List of vehicles that are parking___");
            Vehicle_List.OutList(SystemPark.List_Vehicle_On_Main);
        }
        public void Gua_List_Veh_OFF()
        {
            Console.WriteLine("___List of vehicles that were parking___");
            Vehicle_List.OutList(SystemPark.List_Vehicle_Off_Main);
        }
        public void Gua_Output_Person(Guard person)
        {
            Console.WriteLine("Information of Guard <Person>:");
            Console.WriteLine("         ID           |        Name         |Shifts|  Salary/1day  | Overtime |        Address        |       Call Number     |");
            person.Gua_OutPut();
        }
        public static Guard String_to_Guard(string line)
        {
            string[] a = line.Split(';');
            return new Guard(a[0], a[1], int.Parse(a[2]), long.Parse(a[3]), TimeSpan.Parse(a[4]), a[5], a[6]);
        }
        //Tìm và Xóa 1 vehicle trong List_Vehicle_On_Main
        public Vehicle Find_Delete_Vehicle(string veh_ID)
        {
            for (int i = 0; i < SystemPark.List_Vehicle_On_Main.Count; i++)
                if (SystemPark.List_Vehicle_On_Main[i].Veh_ID == veh_ID)
                {
                    Vehicle veh = SystemPark.List_Vehicle_On_Main[i];
                    SystemPark.List_Vehicle_On_Main.RemoveAt(i);
                    Console.WriteLine("Successful delete Vehicle in List_Vehicle_On!!");
                    return veh;
                }
            Console.WriteLine("Delete failed!!");
            return new Vehicle_Bike();
        }
    }
}